src/qubex/experiment/experiment.py

```
@@ -1019,10 +1029,89 @@ def calc_control_amplitudes(
+    ) -> Waveform:
+        if type is None:
+            type = "drag" if target in self.drag_pi_pulse else "flattop"
+        if use_hpi:
+            return self.x90(target, type=type).repeated(2)
+        try:
+            if type == "flattop":
+                return self.pi_pulse[target]
+            elif type == "drag":
+                return self.drag_pi_pulse[target]
+        except KeyError:
+            raise ValueError(f"Invalid target: {target}")
```

Ruff (0.8.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)